### PR TITLE
rbd-nbd: fixes and improvements for unmap/detach wait for process terminate

### PR DESCRIPTION
**Pull request by trociny**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by trociny** [comment]
@dillaman It makes it work in my environment (openSUSE Tumbleweed with 5.8.4-1-default kernel) but it still does not work on teuthology for both ubuntu and rhel platforms.

The problem is `pidfd_open`. Reading the article [1] about introducing `pidfd_open` syscall, I had an impression that the syscall is equivalent to just opening `/proc/$pid`. But it is not what I am observing. On both local env and teuthology, if I get pidfd by opening  `/proc/$pid`, the `poll` does not wait for the process to terminate and returns immediately. Only  `pidfd_open` syscall works as expected but it is not available for kernels used on teuthology. 

So right now I just returned the unmap workaround to the test script hoping that eventually all platforms will support `pidfd_open` syscall and it will work. Still I have some suspicion that I may do something wrong when opening `/proc/$pid`...

[1] https://lwn.net/Articles/789023/

**Comment by brauner** [comment]
> @dillaman It makes it work in my environment (openSUSE Tumbleweed with 5.8.4-1-default kernel) but it still does not work on teuthology for both ubuntu and rhel platforms.
> 
> The problem is `pidfd_open`. Reading the article [1] about introducing `pidfd_open` syscall, I had an impression that the syscall is equivalent to just opening `/proc/$pid`. But it is not what I am observing. On both local env and teuthology, if I get pidfd by opening `/proc/$pid`, the `poll` does not wait for the process to terminate and returns immediately. Only `pidfd_open` syscall works as expected but it is not available for kernels used on teuthology.
> 
> So right now I just returned the unmap workaround to the test script hoping that eventually all platforms will support `pidfd_open` syscall and it will work. Still I have some suspicion that I may do something wrong when opening `/proc/$pid`...
> 
> [1] https://lwn.net/Articles/789023/

`pidfd_open()` and `/proc/<pid>` fds can't be used interchangeably. The `pidfd_open()` syscall returns an anonymous inode-based file descriptor so a pidfd is a separate concept from `/proc/<pid>`. That's the short explanation. :)

**Comment by trociny** [comment]
@brauner thanks for the explanaiton :-) Any suggestions how to wait for a non-child to exit if `pidfd_open` syscall is not supported? I can just think of periodically polling with `kill(pid, 0)`, which does not look very nice.

**Comment by trociny** [comment]
@dillaman I think it is in good shape for review now. The teuthology test results [1].

[1] https://pulpito.ceph.com/trociny-2020-11-09_10:47:08-rbd-wip-mgolub-testing-distro-basic-smithi/

**Comment by dillaman** [comment]
jenkins test make check

File: src/tools/rbd_nbd/rbd-nbd.cc [hunk 1]
```diff
@@ -1431,18 +1432,41 @@ static void run_server(Preforker& forker, NBDServer *server, bool netlink_used)
...
+
+  cerr << "rbd-nbd: waiting for process exit timed out" << std::endl;
+  return -ETIMEDOUT;
+}
+
 // Eventually it should be replaced with glibc' pidfd_open
 // when it is widely available.
+
+#ifndef __NR_pidfd_open
+#define __NR_pidfd_open 434   /* System call # on most architectures */
```
Comment (dillaman) [inline src/tools/rbd_nbd/rbd-nbd.cc:1490]: Nit: I still think we should avoid defining this if it doesn't exist. Instead, use `#ifdef __NR_pidfd_open` to create a stub method that just returns ENOSYS.

Comment (trociny) [reply src/tools/rbd_nbd/rbd-nbd.cc:1490]: I just copied this from the man page example [1]. And actually to me it makes sense: if the currently installed unistd.h header does not have the syscall, it does not mean it is not supported by the currently running kernel (or a kernel where rbd-nbd will eventually be running) and vice versa, if it is defined in the header it does not mean the kernel will support it.

So I am in favor to leave it as it is currently, but if my reasoning has not changed your mind, I will update the code.

[1] https://man7.org/linux/man-pages/man2/pidfd_open.2.html

Comment (dillaman) [reply src/tools/rbd_nbd/rbd-nbd.cc:1490]: I personally still don't like it -- unless you perhaps wrapped it w/ a `LINUX` guard. The closest other example we have to this in Ceph is the `__NR_syncfs` syscall, but it doesn't attempt to define it.

**Comment by trociny** [comment]
@dillaman updated

**Review by dillaman — APPROVED** [review]
lgtm

**Pull request merged**